### PR TITLE
Feature/tiny 6011

### PR DESCRIPTION
**Pull request by HAFRMO**
Related Ticket:

Description of Changes:
Added options to the uploader failure function to permit the removal of images that could not be uploaded. Defaults to false.

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)
* [x] License headers added on new files (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

**Comment by lnewson** [comment]
I really don't think a global setting is going to work here. The customers don't want it removed for failed network requests, only when they explicitly tell it to be removed.

**Comment by spocke** [comment]
Hmm then it will be more complex to handle we could possibly throw and exception if rtc is enabled and this option is used.

**Comment by spocke** [comment]
Or maybe console warn since it's in a promise.

**Comment by lnewson** [comment]
A thought, could we have a setting to disable this behaviour that the RTC plugin could turn on?

**Comment by spocke** [comment]
We are already patching over the src change with rtc specific stuff so we could just add such an override here as well.

**Comment by TheSpyder** [comment]
Eventually we'll have to build APIs into RTC for internal features to update the model, for now yes we can just disable this setting in the RTC plugin.

File: modules/tinymce/src/core/main/ts/api/EditorUpload.ts [hunk 1]
```diff
@@ -135,7 +135,18 @@ const EditorUpload = function (editor: Editor): EditorUpload {
             blobCache.removeByUri(image.src);
             replaceImageUriInView(image, uploadInfo.url);
           } else if (uploadInfo.error) {
-            ErrorReporter.uploadError(editor, uploadInfo.error);
+            if (uploadInfo.error.options.remove) {
+              if (editor.hasPlugin('rtc')) {
```
Comment (TheSpyder) [inline modules/tinymce/src/core/main/ts/api/EditorUpload.ts:139]: We have `Rtc.isRtc(editor)` as a central check for this 👍 

I didn't realise it was a dynamic option returned by the handler, this is certainly easier than overriding the entire upload handler in the plugin.

Comment (HAFRMO) [reply modules/tinymce/src/core/main/ts/api/EditorUpload.ts:139]: We seem to use isRtc and hasPlugin an equal amount of times ( three ) in the code. Is there a notable difference between the two?

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/EditorUpload.ts:139]: `Rtc.isRtc()` is only available in core and is the preferred method. The `hasPlugin` variant is needed for other plugins, as it isn't part of core.

As a side note, I didn't consider forking here, but that does seem like a better solution for now. Nice work!

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/EditorUpload.ts:139]: Worth noting that the existing `hasRtc` check could be improved to use the new hasPlugin API. eg:
```
export const isRtc = (editor: Editor) => editor.hasPlugin('rtc', true);
```

File: modules/tinymce/src/core/main/ts/file/Uploader.ts [hunk 2]
```diff
@@ -31,13 +31,20 @@ import Tools from '../api/util/Tools';
...
+export type UploadHandler = (blobInfo: BlobInfo, success: (url: string) => void, failure: (err: string, options?: UploadFailureOptions) => void, progress?: (percent: number) => void) => void;
 
 export interface UploadResult {
   url: string;
   blobInfo: BlobInfo;
   status: boolean;
-  error?: string;
+  error?: {
+    options: UploadFailureOptions;
+    error: string;
```
Comment (spocke) [inline modules/tinymce/src/core/main/ts/file/Uploader.ts:46]: Maybe this should be message since result.error.error looks a bit funny, message would also be a bit more similar to the js Error interface.

Comment (HAFRMO) [reply modules/tinymce/src/core/main/ts/file/Uploader.ts:46]: That makes sense.

File: modules/tinymce/src/core/main/ts/api/EditorUpload.ts [hunk 3]
```diff
@@ -135,7 +136,19 @@ const EditorUpload = function (editor: Editor): EditorUpload {
             blobCache.removeByUri(image.src);
             replaceImageUriInView(image, uploadInfo.url);
           } else if (uploadInfo.error) {
-            ErrorReporter.uploadError(editor, uploadInfo.error);
+            if (uploadInfo.error.options.remove) {
+              if (Rtc.isRtc(editor)) {
+                // To be replaced by RTC API to mirror DOM changes when such is implemented.
+                console.log('Could not erase image due to RTC plugin'); // eslint-disable-line no-console
+              } else {
+                editor.undoManager.transact(() => {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/EditorUpload.ts:144]: As discussed on slack, we don't want a new undo level for every failure, only one at the end. So this needs to be changed so that it does something like `editor.undoManager.add()` once the upload of all files is complete.

File: modules/tinymce/src/core/main/ts/api/EditorUpload.ts [hunk 4]
```diff
@@ -135,7 +136,19 @@ const EditorUpload = function (editor: Editor): EditorUpload {
             blobCache.removeByUri(image.src);
             replaceImageUriInView(image, uploadInfo.url);
           } else if (uploadInfo.error) {
-            ErrorReporter.uploadError(editor, uploadInfo.error);
+            if (uploadInfo.error.options.remove) {
+              if (Rtc.isRtc(editor)) {
+                // To be replaced by RTC API to mirror DOM changes when such is implemented.
+                console.log('Could not erase image due to RTC plugin'); // eslint-disable-line no-console
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/EditorUpload.ts:142]: Instead of this, how about rewording this to be something like this?
```
Removing images on failed uploads is currently unsupported for RTC
```
It just clarifies it's just an interim thing, rather than sounding like it'll never work.

File: modules/tinymce/src/core/test/ts/browser/EditorUploadTest.ts [hunk 5]
```diff
@@ -25,174 +23,167 @@ UnitTest.asynctest('browser.tinymce.core.EditorUploadTest', (success, failure) =
...
   };
 
-  const hasBlobAsSource = function (elm) {
-    return elm.src.indexOf('blob:') === 0;
-  };
+  const hasBlobAsSource = (elm) =>
+    elm.src.indexOf('blob:') === 0;
 
-  suite.asyncTest('_scanForImages', function (editor, done, die) {
+  const sScanForImages = (editor) => Log.step('TBA', '_scanForImages', Step.async((done, die) => {
```
Comment (lnewson) [inline modules/tinymce/src/core/test/ts/browser/EditorUploadTest.ts:63]: Types for `editor` here please and all the place below (this wasn't needed previously as it was inferred from `suite.asyncTest`)

File: modules/tinymce/src/core/test/ts/browser/EditorUploadTest.ts [hunk 6]
```diff
@@ -285,67 +276,99 @@ UnitTest.asynctest('browser.tinymce.core.EditorUploadTest', (success, failure) =
       editor.uploadImages(uploadDone),
       editor.uploadImages(uploadDone)
     ]).then(done, die);
-  });
+  }));
+
+  const sUploadConcurrentImagesFailWithRemove = (editor) => Log.step('TBA', 'uploadConcurrentImages fails, with remove', Step.async((done, die) => {
+    let uploadCount = 0, callCount = 0;
+
+    const uploadDone = (result: UploadResult[]) => {
```
Comment (lnewson) [inline modules/tinymce/src/core/test/ts/browser/EditorUploadTest.ts:284]: This really should check the undo stack count at least to make sure it's what we expected. It'd also be good to have 2 or more images that are removed to make sure only a single undo level is added.

File: modules/tinymce/changelog.txt [hunk 7]
```diff
@@ -1,6 +1,7 @@
 Version 5.5.0 (TBD)
     Added new `table_column_resizing` setting to control how the table is affected when resizing columns using the resize bars #TINY-6001
     Added ability to load TinyMCE in inline mode within a ShadowRoot #TINY-6128
+    Added the image `images_upload_handler` can now chose to have images removed on a failed upload #TINY-6011
```
Comment (lnewson) [inline modules/tinymce/changelog.txt:4]: This doesn't read very well, maybe something like this instead?
```
Added the ability to remove images on a failed upload using the `images_upload_handler` failure callback #TINY-6011
```

File: modules/tinymce/src/core/main/ts/api/EditorUpload.ts (hunk 4 above)
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/EditorUpload.ts:142]: Also should this be a `console.warning` or `console.error`? `console.log` is unlikely to be the logging level we want here.

File: modules/tinymce/src/core/main/ts/api/EditorUpload.ts [hunk 8]
```diff
@@ -146,6 +153,20 @@ const EditorUpload = function (editor: Editor): EditorUpload {
           };
         });
 
+        if (imagesToRemove.length > 0) {
+          if (Rtc.isRtc(editor)) {
+            // To be replaced by RTC API to mirror DOM changes when such is implemented.
+            console.error('Removing images on failed uploads is currently unsupported for RTC'); // eslint-disable-line no-console
+          } else {
+            editor.undoManager.transact(() => {
+              Arr.each(imagesToRemove, (element: HTMLImageElement) => {
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/EditorUpload.ts:162]: You don't need the type here, it's already inferred from the type on `imagesToRemove`, so can you please remove it?

Comment (HAFRMO) [reply modules/tinymce/src/core/main/ts/api/EditorUpload.ts:162]: If you insist.
